### PR TITLE
[DI] Add missing deprecation when fetching private services from ContainerBuilder

### PR DESCRIPTION
**Pull request by nicolas-grekas**
| Q             | A
| ------------- | ---
| Branch?       | 3.4
| Bug fix?      | yes
| New feature?  | no
| BC breaks?    | no
| Deprecations? | no
| Tests pass?   | yes
| Fixed tickets | #25242
| License       | MIT
| Doc PR        | -

### CODE REVIEW

**Review by stof — APPROVED** [review]

**Review by chalasr — APPROVED** [review]

**Review by fabpot — APPROVED** [review]

**Comment by fabpot** [comment]
Tests are red though, and this seems related.

**Review by xabbuh — APPROVED** [review]

File: src/Symfony/Component/DependencyInjection/ContainerBuilder.php [hunk 1]
```diff
@@ -776,6 +769,12 @@ public function compile(/*$resolveEnvPlaceholders = false*/)
         }
 
         parent::compile();
+
+        foreach ($this->definitions + $this->aliasDefinitions as $id => $definition) {
+            if (!$definition->isPublic() || $definition->isPrivate()) {
+                $this->removedIds[$id] = true;
```
Comment (nicolas-grekas) [inline src/Symfony/Component/DependencyInjection/ContainerBuilder.php]: This is for ppl doing that, which is a no-op, but one we catch now:
image

File: src/Symfony/Component/DependencyInjection/Tests/ContainerBuilderTest.php [hunk 2]
```diff
@@ -733,6 +733,7 @@ public function testEnvInId()
             PsrContainerInterface::class => true,
             ContainerInterface::class => true,
             'baz_%env(BAR)%' => true,
+            'bar_%env(BAR)%' => true,
```
Comment (nicolas-grekas) [inline src/Symfony/Component/DependencyInjection/Tests/ContainerBuilderTest.php]: the added entries are the same as on 4.0/master: we now list the public-by-private legacy services as removed, even if they're not really for BC in 3.4.

File: src/Symfony/Component/HttpKernel/Controller/ContainerControllerResolver.php [hunk 3]
```diff
@@ -94,7 +94,7 @@ protected function instantiateController($class)
         } catch (\TypeError $e) {
         }
 
-        if ($this->container instanceof Container && in_array($class, $this->container->getRemovedIds(), true)) {
+        if ($this->container instanceof Container && isset($this->container->getRemovedIds()[strtolower($class)])) {
```
Comment (nicolas-grekas) [inline src/Symfony/Component/HttpKernel/Controller/ContainerControllerResolver.php]: that's a bit unrelated, but still found while playing with getRemovedIds: removed services are provided as *keys*

**Comment by nicolas-grekas** [comment]
Now green

**Review by stof — APPROVED** [review]

File: src/Symfony/Component/DependencyInjection/Tests/ContainerBuilderTest.php (hunk 2 above)
Comment (chalasr) [reply src/Symfony/Component/DependencyInjection/Tests/ContainerBuilderTest.php]: won't they appear as removed even if they are not in 4.0?

File: src/Symfony/Component/HttpKernel/Controller/ContainerControllerResolver.php (hunk 3 above)
Comment (stof) [reply src/Symfony/Component/HttpKernel/Controller/ContainerControllerResolver.php]: are you sure about the strtolower ? The normalized id is not the lowercase one

**Review by fabpot — APPROVED** [review]

File: src/Symfony/Component/HttpKernel/Controller/ContainerControllerResolver.php (hunk 3 above)
Comment (nicolas-grekas) [reply src/Symfony/Component/HttpKernel/Controller/ContainerControllerResolver.php]: indeed, fixed

File: src/Symfony/Component/DependencyInjection/Tests/ContainerBuilderTest.php (hunk 2 above)
Comment (nicolas-grekas) [reply src/Symfony/Component/DependencyInjection/Tests/ContainerBuilderTest.php]: wdyt? they *are* in 4.0

Comment (chalasr) [reply src/Symfony/Component/DependencyInjection/Tests/ContainerBuilderTest.php]: definition isn't removed if injected twice or more, is it? If it isn't, wouldn't `getRemovedIds()` return an id that has not been removed?

**Review by xabbuh — APPROVED** [review]

**Review by Simperfit — APPROVED** [review]

**Comment by fabpot** [comment]
Thank you @nicolas-grekas.

**Pull request merged**